tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,747 @@
+
+                    // To check, it's sufficient to see if cached_when and
+                    // true_when are in sync. cached_when will always reflect
+                    // the last value written by the driver. This is because the
+                    // waker register_by_ref's implicit fence (above) ensures
+                    // that any cached_when writes are visible side-effects
+                    // here.
+                    if unsafe { this.inner().cached_when() } == this.inner().true_when() {
+                        break Ok(()).into();
+                    } else {
+                        // We're unsure of the state of this timer, so
+                        // re-register it, and loop again to figure out what's
+                        // up.
+                        unsafe {
+                            this.driver
+                                .reregister(NonNull::new(this.inner.get()).unwrap())
+                        }
+                    }
+                }
+                EntryState::Cancelled => break Ok(()).into(),
+                EntryState::Error(e) => break Err(e.into()).into(),
             }
+        }
+    }
+}
 
-            let actual = entry.state.compare_and_swap(curr, next, SeqCst);
+impl TimerHandle {
+    pub(super) unsafe fn cached_when(&self) -> u64 {
+        unsafe { self.inner.as_ref().cached_when() }
+    }
 
-            if curr == actual {
-                break;
-            }
+    pub(super) unsafe fn sync_when(&self) -> u64 {
+        unsafe { self.inner.as_ref().sync_when() }
+    }
 
-            curr = actual;
-        }
+    pub(super) unsafe fn is_pre_registration(&self) -> bool {
+        let state = unsafe { self.inner.as_ref().state.get(Ordering::Relaxed) };
 
-        // If the state has transitioned to 'elapsed' then wake the task as
-        // this entry is ready to be polled.
-        if !is_elapsed(curr) && is_elapsed(next) {
-            entry.waker.wake();
-        }
+        state == EntryState::NotRegistered
+    }
+
+    /// Tries to set this entry to registered state.
+    ///
+    /// Returns true if successful, or false if already in an error state.
+    ///
+    /// SAFETY: The caller must ensure that the handle remains valid and that
+    /// the state transition rules are not violated.
 
-        // The driver tracks all non-elapsed entries; notify the driver that it
-        // should update its state for this entry unless the entry had already
-        // elapsed and remains elapsed.
-        if !is_elapsed(curr) || !is_elapsed(next) {
-            let _ = inner.queue(entry);
+    pub(super) unsafe fn set_registered(&self) -> bool {
+        unsafe {
+            match self.inner.as_ref().state.get(Ordering::Relaxed) {
+                EntryState::Error(_) => return false,
+                EntryState::Cancelled => panic!("Trying to register a cancelled timer"),
+                EntryState::FiringInProgress => panic!("Unexpected FiringInProgress state"),
```

Comment:
nit: should these be `unreachable`? if these indicate that internal invariants are violated, I think we should prefer `unreachable!` to indicate to a user who sees them that it's definitely not their fault and they should report a bug...
```suggestion
                EntryState::Cancelled => unreachable!("Trying to register a cancelled timer"),
                EntryState::FiringInProgress => unreachable!("Unexpected FiringInProgress state"),
```